progman.c

```diff
@@ -0,0 +1,212 @@
+#include "gqueue.h"
+
+#include "system.h"
+#include "report.h"
+#include "serial.h"
+#include "motion_control.h"
+
+#define PROGRAM_MAX_SIZE (1024 * 3)
+
+DECLARE_QUEUE(prog_buf, uint8_t, PROGRAM_MAX_SIZE);
+
+enum progman_state {
+  E_PROGMAN_CONSUMING = 0,
+  E_PROGMAN_COMMENT
+};
+
+static struct {
+  uint8_t line_assy[LINE_BUFFER_SIZE];
+  uint8_t line_pos;
+  bool in_comment;
+  enum progman_state state;
+} progman;
+
+void progman_init(void)
+{
+  progman.line_pos = 0;
+
+  queue_init(&prog_buf, sizeof(uint8_t), PROGRAM_MAX_SIZE);
+
+  while (!queue_is_empty(&prog_buf)) {
+    uint8_t nil;
+    queue_dequeue(&prog_buf, &nil);
+  }
+  progman.state = E_PROGMAN_CONSUMING;
+}
+
+bool progman_handle_runtime_cmd(char data)
+{
+  switch (data) {
+  case CMD_COUNTER_REPORT:
+    request_report(REQUEST_COUNTER_REPORT,0);
+    break;
+  case CMD_VOLTAGE_REPORT:
+    request_report(REQUEST_VOLTAGE_REPORT,0);
+    break;
+  case CMD_STATUS_REPORT:
+    request_report(REQUEST_STATUS_REPORT,0);
+    break;
+  case CMD_LIMIT_REPORT:
+    request_report(REQUEST_LIMIT_REPORT,0);
+    break;
+  case CMD_CYCLE_START:
+    /* Set as true */
+    SYS_EXEC |= EXEC_CYCLE_START;
+    break;
+  case CMD_FEED_HOLD:
+    /* Set as true */
+    SYS_EXEC |= EXEC_FEED_HOLD;
+    break;
+  case CMD_RESET:
+    /* Call motion control reset routine. */
+    mc_reset();
+    break;
+  case '\n':			/* Intentional fall through */
+  case '\r':
+    /* If we got a bare newline character out of context, it's likely
+       due to the fact that our previous character was consumed as
+       a runtime command, so we should just swallow it */
+    if ((progman.line_pos < 1) &&
+	(E_PROGMAN_CONSUMING == progman.state)) {
+      report_status_message(STATUS_OK);
+      break;
+    }
+  default:
+    /* If we didn't handle the command, return false  */
+    return false;
+  }
+
+  return true;
+}
+
+static bool progman_consume_byte(char data)
+{
+
+  /* Immediately try to toss out any comments */
+  switch (data) {
+  case '(':
```

Comment:
Is a nested comment valid in the GCODE standard?

Something like: G99 
```(comments (more comments) done with comments)```
would cause a problem, no?